{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#@title Intro & Install\n",
    "\n",
    "#@markdown # Aims\n",
    "#@markdown This notebook takes you through the basics of molecular mechanics,\n",
    "#@markdown as a primer for docking.\n",
    "\n",
    "#@markdown # Step 1: Press play!\n",
    "\n",
    "#@markdown ### Preface on Colab\n",
    "\n",
    "#@markdown ‚Üê Press the button with the play icon to run a _cell_.\n",
    "#@markdown In this cell: install requirements.\n",
    "\n",
    "#@markdown This is a _Colab notebook_, a variant of a Jupyter notebook.\n",
    "#@markdown If you are not in Colab press [this](https://colab.research.google.com/github/matteoferla/Fragment-hit-follow-up-chemistry/blob/main/colab/upload_prep.ipynb).\n",
    "\n",
    "#@markdown Colab runs in Google's servers, hence why you will get asked\n",
    "#@markdown to sign in if not done so already.\n",
    "#@markdown Likewise it will ask if you trust the author (Matteo Ferla),\n",
    "#@markdown if unsure about whether you should trust anything I do\n",
    "#@markdown [click here for details](https://www.youtube.com/watch?v=dQw4w9WgXcQ).\n",
    "\n",
    "#@markdown To inspect code press `show code` ‚Üì\n",
    "\n",
    "#@markdown Still confused about notebook? Then see the [about_notebooks.md page in the repo](https://github.com/matteoferla/DTC-compchem-practical/blob/main/about_notebooks.md).\n",
    "\n",
    "#@markdown You will NOT get a prize for pressing `Run all` and going for a coffee.\n",
    "#@markdown Read the code, google up and use `help(ü§ñ)` where ü§ñ is the object.\n",
    "#@markdown (No, you cannot use emoji as variables in vanilla Python).\n",
    "#@markdown In this notebook, the imported module `dtc` will have `dtc.show_source(ü§ñ)` will show the source code\n",
    "#@markdown in a colourful way (it calls `inspect.getsource` a handy function). Please use it.\n",
    "#@markdown Somethings are hidden in the module within the repo, you can pretend they work by magic\n",
    "\n",
    "#@markdown # Step 2: Basic concepts\n",
    "#@markdown While we wait N minutes for conda to install things,\n",
    "#@markdown we will go through the basics of molecular mechanics in the [theory.md page in the repo](https://github.com/matteoferla/DTC-compchem-practical/blob/main/theory.md).\n",
    "#@markdown ![mm](https://upload.wikimedia.org/wikipedia/commons/thumb/5/5c/MM_PEF.png/1920px-MM_PEF.png)\n",
    "\n",
    "#@markdown If you do not want to read anything (one of those days, eh?),\n",
    "#@markdown then watch the [video in the repo](https://github.com/matteoferla/DTC-compchem-practical/raw/main/media/stefans_animation.mp4) that shows the fragment hits\n",
    "#@markdown from the Mac1 domain of NSP13 protein of SARS-COV-2 screens that were expanded into strong binders!\n",
    "#@markdown In the next cell there are questions to answer based on the video though!\n",
    "\n",
    "# ------------------------------------------------\n",
    "\n",
    "import os\n",
    "\n",
    "#source /etc/os-release && echo $PRETTY_NAME\n",
    "with open('/etc/os-release') as fh:\n",
    "  for line in fh:\n",
    "    if line and line[0] != '#':\n",
    "      os.environ[line.split('=')[0]] = line.split('=')[1]\n",
    "print(f'Running {os.environ[\"PRETTY_NAME\"]}') # on {os.environ[\"HOST\"]} as {os.environ[\"USER\"]}')\n",
    "\n",
    "# ------------------------------------------------\n",
    "\n",
    "#@title Installation\n",
    "local_debug = False\n",
    "if local_debug:\n",
    "    raise Exception('CURRENTLY IN DEBUG MODE.... REMEMBER TO CLEAR ALL CELLS!')\n",
    "\n",
    "## Install all requirements and get some goodies\n",
    "!pip install git+https://github.com/matteoferla/DTC-compchem-practical.git\n",
    "# this will be called as:\n",
    "# import DTC_compchem_practical as dtc\n",
    "\n",
    "## Jupyter lab? use `trident-chemwidgets`\n",
    "!pip install git+https://github.com/matteoferla/JSME_notebook_hack.git\n",
    "!pip install --upgrade plotly matplotlib\n",
    "\n",
    "from google.colab import output  # noqa (It's a colaboratory specific repo)\n",
    "output.enable_custom_widget_manager()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "#@title Video summary\n",
    "#@markdown This video (press play to run), from the Mac1 domain of NSP13 protein of SARS-COV-2 screens summarises perfectly\n",
    "#@markdown the challenges of a fragment based drug discovery screen.\n",
    "\n",
    "from IPython.display import HTML, display\n",
    "display(\n",
    "    HTML(\"\"\"\n",
    "        <video alt=\"test\" controls>\n",
    "            <source src=\"https://github.com/matteoferla/DTC-compchem-practical/raw/main/media/stefans_animation.mp4\" type=\"video/mp4\">\n",
    "        </video>\n",
    "    \"\"\")\n",
    ")"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Questions\n",
    "So you chose not to read the theory page, eh? Well, let's see if you can answer these questions based on the video.\n",
    "\n",
    "Please replace alien-monster emoji with your answers. Remember to save or pdf-print your notebook before leaving\n",
    "to show your grandkids.\n",
    "\n",
    "> This is an enzyme, which binds to a nucleotide modification on protein. Can you guess where its native substrate binds? Which fragment site would be good for a competitive inhibitor?\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> Do all fragments bind to the protein with the same out of contacts per atom? If not, what is ligand efficiency?\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    ">  Is the protein rigid? I.e. lock & key Vs. induced fit.\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> Towards the end a loop \"everts\", reflecting a native product bound state vs. a native substrate bound state. Would repacking sidechains model this change?\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> Would you reckon, the apo structure choice affect docking results?\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> Some compounds have labels like `ZINC922`. What is Zinc?\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> The theoretical merger has a ester bond, while the analogue in make-on-demand space has an amide. Can you guess why?\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> What makes a lead, \"drug-like\"? (cf. [Lipinski rule_of 5](https://en.wikipedia.org/wiki/Lipinski%27s_rule_of_five) ) And is a carboxyl group good? What is a bioisostere?\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> Is synthetic accessibility really important, why?\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> Is chirality easy to make with _synthetic chemistry_ (not biocatalysis)?\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> Different isomers of racemic compounds bind in different poses within the same crystal. Do these ligands both bind together in the same protein macromolecule at the same time, sequentially in time or on different macromolecules in the crystal lattice?\n",
    "\n",
    "üëæüëæüëæ"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Step 3: Playing with molecules in RDKit\n",
    "\n",
    "Let's play with a molecule. We will use RDKit, a cheminformatics library.\n",
    "It is the most popular cheminformatics library in Python.\n",
    "\n",
    "Go to Wikipedia and search for a molecules and copy its SMILES from the infobox.\n",
    "\n",
    "If you want a top searched molecule from Wikipedia, [here is an analysis](https://github.com/matteoferla/Wikipedian-compounds) with SMILES in the csv file. For SMILES of common placeholders see the [last section of this blog post](https://www.blopig.com/blog/2023/08/placeholder-compounds-distraction-vs-accuracy/)."
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "#@markdown Input name and SMILES of molecule\n",
    "mol_name: str =  'üëæüëæüëæ'   #@param {type:\"string\"}\n",
    "smiles: str =  'üëæüëæüëæ'   #@param {type:\"string\"}\n",
    "if local_debug:\n",
    "    mol_name: str =  'caffeine'   #@param {type:\"string\"}\n",
    "    smiles: str =  'CN1C=NC2=C1C(=O)N(C(=O)N2C)C'   #@param {type:\"string\"}"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "## Let's load it into RDKit and display it\n",
    "# paying attention to the code\n",
    "\n",
    "from IPython.display import display\n",
    "from rdkit import Chem\n",
    "\n",
    "#RDKit can and will misbehave in Colab unless this line is called to activate it\n",
    "from rdkit.Chem.Draw import IPythonConsole\n",
    "\n",
    "mol: Chem.Mol = Chem.MolFromSmiles(smiles)\n",
    "mol.SetProp('_Name', mol_name)\n",
    "display(mol)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "#@markdown Let's look at its partial charges (Gasteiger-Marsili).\n",
    "#@markdown Partial charges are a way to model the electron density of a molecule.\n",
    "#@markdown In molecular mechanics, as used in docking, they are generally precomputed before the run during parameterisation.\n",
    "\n",
    "from rdkit.Chem.Draw import SimilarityMaps\n",
    "from rdkit.Chem import AllChem, Draw\n",
    "from PIL import Image\n",
    "from io import BytesIO\n",
    "#@markdown via `rdkit.Chem.AllChem.ComputeGasteigerCharges`\n",
    "AllChem.ComputeGasteigerCharges(mol)\n",
    "contribs = [a.GetDoubleProp('_GasteigerCharge') for a in mol.GetAtoms()]\n",
    "# old way\n",
    "# fig = SimilarityMaps.GetSimilarityMapFromWeights(mol, contribs, colorMap='jet', contourLines=10)\n",
    "d = Draw.MolDraw2DCairo(400, 400)\n",
    "SimilarityMaps.GetSimilarityMapFromWeights(mol, contribs, draw2d=d)\n",
    "d.FinishDrawing()\n",
    "Image.open( BytesIO(d.GetDrawingText()) )"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Questions\n",
    "\n",
    "> what is 'RDKit'?\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> What is a SMILES (or 'SMILEString')?\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> What does the function `AllChem.ComputeGasteigerCharges` do?\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> In addition to Marsili-Gasteiger partial charges, there is another form of partial charges, what is it?\n",
    "\n",
    "üëæüëæüëæ"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "#@markdown As a rewarded for acing those questions. Here is a fun function:\n",
    "#@markdown adding meme gifs to your outputs, because ... ehr... science?\n",
    "#@markdown Ask a demonstrator to reward you with the password to unlock the fun!\n",
    "#@markdown Please do not tell the serious people of this cell or Matteo will get in trouble.\n",
    "\n",
    "password = 'üëæüëæüëæ' #@param {type:\"string\"}\n",
    "token = b'gAAAAABjhNPjteatHENaWRl63ddoEY6VRJCegYVx1ROsLwDi-4NOIXJ3gr8pYffO42OGVzJDlr0qg_F_1Sbh1RP9W5jzyyW12diw58RDj2ydR96o0kIJVPGbl7JMWJy0BjOiowGnvPET'\n",
    "import DTC_compchem_practical as dtc\n",
    "import os\n",
    "os.environ['GIPHY_API'] = dtc.decrypt(token, password)\n",
    "\n",
    "from gist_import import GistImporter\n",
    "\n",
    "retrieve_giphy = GistImporter('25a53d54d3ba65919610bb34b188ac67')['retrieve_giphy']\n",
    "\n",
    "retrieve_giphy('protein')"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "#@markdown Let's alter the molecule and see how the partial charges change.\n",
    "from jsme_hack.rdkit import JSMERDKit\n",
    "import ipywidgets as widgets\n",
    "\n",
    "button = widgets.Button(description=\"Calculate\", icon='fire')\n",
    "output = widgets.Output(layout={'border': '1px solid black'})\n",
    "display(button, output)\n",
    "jsme = JSMERDKit(mol)\n",
    "def on_click(remove:bool):\n",
    "    alt_mol = jsme.mol\n",
    "    AllChem.ComputeGasteigerCharges(alt_mol)\n",
    "    contribs = [a.GetDoubleProp('_GasteigerCharge') for a in alt_mol.GetAtoms()]\n",
    "    with output:\n",
    "        d = Draw.MolDraw2DCairo(400, 400)\n",
    "        SimilarityMaps.GetSimilarityMapFromWeights(mol, contribs, draw2d=d)\n",
    "        d.FinishDrawing()\n",
    "        Image.open( BytesIO(d.GetDrawingText()) )\n",
    "\n",
    "button.on_click(on_click)\n",
    "on_click(False)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "#@title Connectivity\n",
    "#@markdown A molecule is represented as a graph. The atoms are the nodes and the bonds are the edges.\n",
    "#@markdown How is your molecule connected?\n",
    "\n",
    "from typing import Tuple, List, Iterator\n",
    "import operator\n",
    "\n",
    "for i, atom in enumerate(mol.GetAtoms()):  #: Tuple[int, Chem.Atom]\n",
    "    # neighbours are atoms it is bonded to\n",
    "    neigh: Tuple[Chem.Atom] = atom.GetNeighbors()\n",
    "    neigh_i: Tuple[int] = tuple(map(Chem.Atom.GetIdx, neigh))\n",
    "    bonds: Tuple[Chem.Bond] = (mol.GetBondBetweenAtoms(i, n) for n in neigh_i)\n",
    "    # bondtype is an enum `Chem.Bond.GetBondType().name`\n",
    "    bondtypes: Iterator[Chem.BondType] = map(Chem.Bond.GetBondType, bonds)\n",
    "    bondtype_names: Tuple[str] = tuple(map(operator.attrgetter('name'), bondtypes))\n",
    "    print(f'{atom.GetSymbol()} atom (index {i}) is connected to {neigh_i} via {bondtype_names}')"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "#@title Conformer generation\n",
    "#@markdown A molecule is not only its connectivity, but also its 3D shape, the conformer.\n",
    "#@markdown Each node has an associated 3-dimensional vector in Cartesian space.\n",
    "\n",
    "from rdkit.Chem import rdDistGeom, rdMolAlign, Draw\n",
    "\n",
    "# Add Hs\n",
    "hydromol = AllChem.AddHs(mol)\n",
    "\n",
    "# Generate a single conformer\n",
    "mol3d = Chem.Mol(hydromol) # making a copy\n",
    "AllChem.EmbedMolecule(mol3d)\n",
    "mol3d.SetProp('_Name', 'single-conf')\n",
    "\n",
    "# Generate a 2D representation\n",
    "mol2d = Chem.Mol(hydromol) # making a copy\n",
    "AllChem.Compute2DCoords(mol2d)\n",
    "mol2d.SetProp('_Name', '2D-repr')\n",
    "\n",
    "# Generate multiple unique conformers\n",
    "multimol = Chem.Mol(hydromol)\n",
    "numconf: int = 10\n",
    "param = rdDistGeom.ETKDGv2()\n",
    "param.pruneRmsThresh = 0.1\n",
    "cids = rdDistGeom.EmbedMultipleConfs(multimol, numconf, param)\n",
    "mp = AllChem.MMFFGetMoleculeProperties(multimol, mmffVariant='MMFF94s')\n",
    "AllChem.MMFFOptimizeMoleculeConfs(multimol, numThreads=0, mmffVariant='MMFF94s')\n",
    "multimol.SetProp('_Name', 'multi-conf')\n",
    "print(f'{multimol.GetNumConformers()} conformers made out of {numconf}')\n",
    "\n",
    "mols = [mol2d, mol3d]\n",
    "for i, conf in enumerate(multimol.GetConformers()):\n",
    "    m = Chem.Mol(hydromol)\n",
    "    m.AddConformer(conf)\n",
    "    m.SetProp('_Name', f'conf_{i}')\n",
    "    mols.append(m)\n",
    "display(Draw.MolsToGridImage(mols,\n",
    "                             legends=[mol.GetProp('_Name') if mol.HasProp('_Name') else '-' for mol in mols],\n",
    "                             subImgSize=(200,200), useSVG=True,\n",
    "                             molsPerRow=4))"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "#@title Test: distort molecule and Constrained MMFF\n",
    "import nglview as nv\n",
    "from rdkit.Chem import rdMolAlign\n",
    "from io import StringIO\n",
    "from gist_import import GistImporter\n",
    "from typing import Callable\n",
    "align_mols: Callable[[Chem.Mol, Chem.Mol], None]= GistImporter('e9399caa07a9206d05330b0c5aca6ec1')['align_mols']\n",
    "\n",
    "\n",
    "from rdkit.Geometry import Point3D\n",
    "\n",
    "def shift_atom(mol: Chem.Mol, atom_idx:0, x_offset: float=0, y_offset: float=0, z_offset: float=0):\n",
    "    \"\"\"\n",
    "    Shifts atom indexed ``atom_idx`` in ``conf`` by ``(atom_idx)``\n",
    "    \"\"\"\n",
    "    conf: Chem.Conformer = mol.GetConformer()\n",
    "    p: Point3D = conf.GetAtomPosition(atom_idx)\n",
    "    # let's shift it by 2√Ö on atom 0 on axis x\n",
    "    new_p: Point3D = Point3D(p.x + x_offset, p.y + y_offset, p.z + z_offset)\n",
    "    conf.SetAtomPosition(atom_idx, new_p)\n",
    "    mol.GetAtomWithIdx(atom_idx).SetBoolProp('Fixed', True)\n",
    "\n",
    "# make 3D\n",
    "mol2 = AllChem.AddHs(mol)\n",
    "AllChem.EmbedMolecule(mol2)\n",
    "## ----------------------------------------------------------------------------\n",
    "\n",
    "#@markdown Run this a few times and change the constraints!\n",
    "# Please tinker with these values:\n",
    "shift_atom(mol2, atom_idx=0, x_offset=2, z_offset=5)\n",
    "shift_atom(mol2, atom_idx=1, y_offset=-1)\n",
    "mol2.GetAtomWithIdx(3).SetBoolProp('Fixed', True)\n",
    "# in the next block `AllChem.MMFFGetMoleculeForceField.MMFFAddPositionConstraint` will be called\n",
    "# constrain these.\n",
    "## ----------------------------------------------------------------------------\n",
    "\n",
    "Chem.SanitizeMol(mol2)\n",
    "\n",
    "p = AllChem.MMFFGetMoleculeProperties(mol2, 'MMFF94')\n",
    "if p is None:\n",
    "    raise ValueError(f'MMFF cannot work on a molecule that has errors!')\n",
    "\n",
    "ff = AllChem.MMFFGetMoleculeForceField(mol2, p)\n",
    "# restrain\n",
    "for atom in mol2.GetAtomsMatchingQuery(Chem.rdqueries.HasPropQueryAtom('Fixed', negate=False)):\n",
    "    # Atom cannot move beyond 2 Aangstroems\n",
    "    ff.MMFFAddPositionConstraint( atom.GetIdx(), 2, 1e5)\n",
    "pre: float = ff.CalcEnergy()\n",
    "outcomes = {-1: 'MMFF Minisation could not be started',\n",
    "            0: 'MMFF Minisation was successful',\n",
    "            1: 'MMFF Minisation was run, but the minimisation was not unsuccessful'}\n",
    "try:\n",
    "    m: int = ff.Minimize()\n",
    "    print(outcomes.get(m, \"I√§! I√§! Cthulhu fhtagn! Ph'nglui mglw'nafh Cthulhu R'lyeh wgah'nagl fhtagn\"))\n",
    "except RuntimeError as error:\n",
    "    print(f'MMFF minimisation failed {error.__class__.__name__}: {error}')\n",
    "\n",
    "post: float = ff.CalcEnergy()\n",
    "print(f'‚àÜG went from {pre} to {post} kcal/mol')\n",
    "#rdMolAlign.AlignMol(mol2, mol)\n",
    "#align_mols(mol2, mol)\n",
    "view = nv.show_rdkit(mol2)\n",
    "# fh = StringIO(Chem.MolToPDBBlock(mol))\n",
    "# view.add_component(fh, ext='mol')\n",
    "\n",
    "display(view)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "#@title Get fingerprints\n",
    "#@markdown `AllChem.GetMorganFingerprint` will generate the Morgan fingerprints of a molecule.\n",
    "#@markdown see [Morgan fingerprinting](https://www.rdkit.org/docs/GettingStartedInPython.html#feature-definitions-used-in-the-morgan-fingerprints) for more.\n",
    "#@markdown This is helpful for substructure searching, but also for clustering molecules.\n",
    "#@markdown Chemical space is vast. Catalogue space and synthetically accessible space are a tiny fraction of it.\n",
    "#@markdown So to navigate chemical space, we need to easy compare them.\n",
    "#@markdown In docking, you will cannot simply dock everything. You will need to filter out molecules that are too similar to each other.\n",
    "\n",
    "from rdkit import Chem\n",
    "from rdkit.Chem import AllChem\n",
    "\n",
    "info = {}\n",
    "morgan = AllChem.GetMorganFingerprintAsBitVect(mol,3,nBits=1024,bitInfo=info)\n",
    "#@markdown We now have a series of 0s and 1s (1024 of them) to describe each 3-atom chunk, a fingerprint, in the molecule\n",
    "print('fingerprint bitmap: ', morgan.ToBitString()[:20]+'...'+morgan.ToBitString()[-20:],\n",
    "      'N bits: ', len(morgan.ToBitString()),\n",
    "       'N non-zero elements: ', morgan.GetNumOnBits())\n",
    "print('The substructures are: ', info)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Questions\n",
    "\n",
    "> Why is there a drop in Gibbs free energy (a potential) after minimisation? (Unusure? Next notebook will explain it further)\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> How does this relate to entropy of the system (think chelate effect)? (Unusure? Next notebook will explain it further)\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> What does `AllChem.EmbedMolecule` do?\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> What command does a 2D representation?\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> What is the difference between a `Chem.Mol` instance and its `Chem.Conformer(s)`? And do both store cartesian atomic positions?\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> Thinking of neural networks... Does the order of the atoms matter? Does rototranslating a whole ligand-protein complex change the binding energy? If not, would this transition and order invariance have an effect in certain NNs? Are there two ways a way to avoid this? (hint: Google equivariance)\n",
    "\n",
    "üëæüëæüëæ"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Synthetic accessibility & make-on-demand space\n",
    "Not all compounds can be made easily or are available from a vendor\n",
    "Fragments are not bought like large common chemicals from Sigma, Fischer, Alfa Aesar or German-Merck etc.\n",
    "Instead they are bought in low mg quantities from Enamine, WuXi or ChemDiv among specialist vendors.\n",
    "These can be divided into off-the-shelf or make-on-demand.\n",
    "The former can be search using for example [www.molport.com](https://www.molport.com/shop/find-chemicals).\n",
    "While the latter can be a very large space, the prime example of this is\n",
    "Enamine REAL, which has 5 billion molecules, while their Store has 1.7 million.\n",
    "But even then 5e6 does not cover the whole of possible chemistry. The site [zinc.docking.org](https://zinc.docking.org/substances/home/) collates the former and some of the latter.\n",
    "\n",
    "Go to [sw.docking.org](https://sw.docking.org/) and try drawing a molecule.\n",
    "Also try picking a _natural compound_ from within say [Wiki category: Flavonoids](https://en.wikipedia.org/wiki/Category:Flavonoids)\n",
    "and copy the SMILES from the infobox into Smallworld. Just because it exists does not mean you can buy it.\n",
    "\n",
    "## Questions\n",
    "\n",
    "> What is the SMILES of the compound you searched in SmallWorld? And what was the distance to a purchasable analogue?\n",
    "\n",
    "üëæüëæüëæ\n",
    "\n",
    "> Did you find a molecule that has a Wikipedia page but is not in EnamineREAL DB?\n",
    "\n",
    "üëæüëæüëæ"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Rototranslations\n",
    "Rototranslations are the 3D transformations that can be applied to a molecule.\n",
    "A (4,4) matrix is used to represent them, this encodes all affine transformations,\n",
    "including translations, rotations, reflections, scaling and shearing.\n",
    "The latter three do not count for molecules!\n",
    "Actually this means that we will be doing extra song-and-dance due to operations on this form of matrix, but it is worth it for didactic purposes...\n",
    "We will also use radians to max out on the confusion.\n",
    "\n",
    "Successive steps of rototranslations can be combined by matrix multiplication.\n",
    "To rotate a molecule by 90¬∞ around the z-axis around a point,\n",
    "one would first translate the molecule so that the point is at the origin,\n",
    "then rotate it and then translate it back.\n",
    "This is done by multiplying the matrices together.\n",
    "\n",
    "For theory see this [helpful guide](https://www.euclideanspace.com/maths/geometry/affine/matrix4x4/index.htm).\n",
    "Do note that there are two types of transformations: [active/alibi and passive/alias](https://en.wikipedia.org/wiki/Active_and_passive_transformation) and PyMol, NGL etc. will use one or the other ‚Äîyou might get stuck in the future by this, so hopefully this passing note will help future-you!"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "from rdkit import Chem\n",
    "from rdkit.Chem.Draw import IPythonConsole\n",
    "from rdkit.Chem import AllChem\n",
    "from rdkit import Geometry\n",
    "\n",
    "def transform(conformer: Chem.Conformer, affine_matrix):\n",
    "    \"\"\"\n",
    "    Applies a 4x4 affine matrix to a conformer.\n",
    "    \"\"\"\n",
    "    for i in range(conformer.GetNumAtoms()):\n",
    "        position: Geometry.Point3D = conformer.GetAtomPosition(i)\n",
    "        # we make the 3D vector homogeneous by appending a 1, the homogenous component\n",
    "        # one for position, zero for direction\n",
    "        position_homogeneous: npt.NDArray[np.float64] = np.append(list(position), 1)\n",
    "        transformed: npt.NDArray[np.float64] = np.dot(affine_matrix, position_homogeneous)\n",
    "        # we no longer need the homogenous component!\n",
    "        transformed_position = transformed[:3]\n",
    "        conformer.SetAtomPosition(i, transformed_position)\n",
    "\n",
    "# Test molecule\n",
    "# trans-Anethole is the main \"aroma\" component of anise oil and fennel oil\n",
    "# don't like anise? The pick a different one!\n",
    "original = AllChem.AddHs(Chem.MolFromSmiles('CC=CC1=CC=C(C=C1)OC'))\n",
    "AllChem.EmbedMolecule(original)\n",
    "\n",
    "# easy peasy\n",
    "translation_matrix = np.array([\n",
    "    [1, 0, 0, 1],  # Translation along x-axis by 1 unit\n",
    "    [0, 1, 0, 0],  # No translation along y-axis\n",
    "    [0, 0, 1, 0],  # No translation along z-axis\n",
    "    [0, 0, 0, 1]   # Homogeneous coordinate\n",
    "])\n",
    "\n",
    "mol = Chem.Mol(original)\n",
    "transform(mol.GetConformer(), translation_matrix)\n",
    "\n",
    "import py3Dmol\n",
    "import DTC_compchem_practical as dtc\n",
    "\n",
    "view: py3Dmol.view = dtc.get_mols_view(whiteCarbon=original, yellowCarbon=mol)\n",
    "view.show()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# Rotation\n",
    "# That was not complicated. Let's do a rotation.\n",
    "\n",
    "# first let's do some trigonometry\n",
    "def create_rotation_matrix(axis: Sequence, angle: float):\n",
    "    \"\"\"\n",
    "    Create a 4x4 affine matrix for rotation.\n",
    "\n",
    "    :param axis: The axis of rotation (x, y, z).\n",
    "    :param angle: The angle of rotation in radians.\n",
    "    :return: 4x4 affine transformation matrix.\n",
    "    \"\"\"\n",
    "    # Normalize the axis vector\n",
    "    axis = np.asarray(axis)\n",
    "    axis = axis / np.sqrt(np.dot(axis, axis))\n",
    "    # Compute the cosine and sine of the angle\n",
    "    c = np.cos(angle)\n",
    "    s = np.sin(angle)\n",
    "    C = 1 - c\n",
    "    # Compute the components of the rotation matrix\n",
    "    x, y, z = axis\n",
    "    r11 = x*x*C + c\n",
    "    r12 = x*y*C - z*s\n",
    "    r13 = x*z*C + y*s\n",
    "    r21 = y*x*C + z*s\n",
    "    r22 = y*y*C + c\n",
    "    r23 = y*z*C - x*s\n",
    "    r31 = z*x*C - y*s\n",
    "    r32 = z*y*C + x*s\n",
    "    r33 = z*z*C + c\n",
    "    # Construct the rotation matrix\n",
    "    rotation_matrix = np.array([\n",
    "        [r11, r12, r13, 0],\n",
    "        [r21, r22, r23, 0],\n",
    "        [r31, r32, r33, 0],\n",
    "        [0,   0,   0,   1]\n",
    "    ])\n",
    "    return rotation_matrix\n",
    "\n",
    "def create_translation_matrix(x: float=0, y: float=0, z: float=0):\n",
    "    \"\"\"\n",
    "    Create a 4x4 affine matrix for translations.\n",
    "\n",
    "    :param x: ...\n",
    "    :param y: ...\n",
    "    :param z: ...\n",
    "    :return: 4x4 affine transformation matrix.\n",
    "    \"\"\"\n",
    "    # identity_matrix\n",
    "    v = np.array([x, y, z])\n",
    "    translation_matrix = np.identity(4)\n",
    "    translation_matrix[0:3, 3] = v\n",
    "    return translation_matrix\n",
    "\n",
    "def get_centroid(conf: Chem.Conformer) -> npt.NDArray[np.float64]:\n",
    "    \"\"\"\n",
    "    geometric center of all the atoms' coordinates\n",
    "    Not centre of mass\n",
    "    \"\"\"\n",
    "    return sum(conf.GetPositions()) / conf.GetNumAtoms()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "In the previous cell we defined a few functions:\n",
    "\n",
    "* `cen: npt.NDArray[np.float64] = get_centroid(mol.GetConformer())`\n",
    "* `translation_by_one: npt.NDArray[np.float64] = create_translation_matrix(x=1)`\n",
    "* `rotation_by_90: npt.NDArray[np.float64] = create_rotation_matrix([1,0,0], angle=np.pi / 2)`\n",
    "\n",
    "Let's use them to rotate the molecule by 90¬∞ around the x-axis around a point.\n",
    "As mentioned, we will first translate the molecule so that the point is at the origin,\n",
    "then rotate it and then translate it back.\n",
    "\n",
    "But oh no! Matteo accidentally deleted the cell with the code to do it!\n",
    "[Demonstrators seem answers.md]"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# first let's move the molecule by 5 √Ö along the x-axis\n",
    "transform(original.GetConformer(), create_translation_matrix(x=5) )\n",
    "print(get_centroid(original.GetConformer()))\n",
    "\n",
    "# then let's make a copy and rotate it by 90¬∞ around the x-axis\n",
    "mol = Chem.Mol(original)\n",
    "\n",
    "üëæüëæüëæ\n",
    "üëæüëæüëæ\n",
    "üëæüëæüëæ\n",
    "# combine the affine transform matrices by matrix multiplication\n",
    "rotation_on_spot: npt.NDArray[np.float64] = np.matmul(np.matmul(üëæüëæüëæ,üëæüëæüëæ), üëæüëæüëæ)\n",
    "transform(mol.GetConformer(), rotation_on_spot )\n",
    "\n",
    "dtc.get_mols_view(whiteCarbon=original, yellowCarbon=mol).show()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# ## Questions\n",
    "\n",
    "> What does a translation look like in a 4x4 matrix?\n",
    "üëæüëæüëæ\n",
    "> A dot product between a 3x1 vector and a 4x4 matrix is not possible, what is the missing detail?\n",
    "üëæüëæüëæ\n",
    "> How do you rotate a molecule around a point?\n",
    "üëæüëæüëæ"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## RMSD\n",
    "\n",
    "The root mean square deviation (RMSD) is a measure of the average distance between the atoms of two molecules.\n",
    "There are four things to note:\n",
    "\n",
    "* Do you want the conformers to be superposed first or what it in place? Always be careful of this!\n",
    "* Mean of squares is more or less sensitive to bigger values and say regular arithmetic mean?\n",
    "* Are isomorphism taken into account? (Flip a benzene ring and it is the same molecule bar for atom numbering and naming)\n",
    "* Atom mapping are always important\n",
    "\n",
    "\"align\" --> overlay two sequences (of matched atoms) pairwise\n",
    "\"superpose\" --> overlay with rototranslations\n",
    "\"superimpose\" --> overlay with translations only (i.e. avoid using this word)\n",
    "üëæüëæüëæ"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "print('In place: ', AllChem.CalcRMS(original, mol))\n",
    "print('Superimposed: ', AllChem.GetBestRMS(original, mol))"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Crude docking\n",
    "\n",
    "Using the Mac1 from the video, let's dock them into the protein.\n",
    "Normally a classic testbed is BRD4 but the video was fun.\n",
    "\n",
    "Normally, data polishing does not happen by magic.\n",
    "Template choice is important, generally the native substrate-bound structure stripped the substrate is best.\n",
    "Using the ligand-bound structure removed of the ligand for docking the ligand is very much cheating\n",
    "as you normally don't have that structure...\n",
    "This dataset was created for this [blog post](https://www.blopig.com/blog/2023/11/the-workings-of-fragmensteins-rdkit-neighbour-aware-minimisation/).\n",
    "\n",
    "Gibbs free energy (G) is enthalpy (H) minus temperature times entropy (TS).\n",
    "Enthalpy is internal energy (U) plus pressure times volume (PV).\n",
    "The logarithm of binding as measured in K_D (pKD) is proportional to Gibbs free energy over Boltzmann's constant (k) times temperature (T).\n",
    "Therefore, the internal energy (U) is not the same as K_D and there is entropy (S) to consider.\n",
    "Yes, from this you will have spotted that K_D is temperature dependent, but please don't say it aloud or you will make cheminformatitians cry.\n",
    "\n",
    "Here will simply extract the neighbouring residues in RDKit and \"freeze\" them.\n",
    "A frozen particle in MM parlance is a particle that is not allowed to move.\n",
    "Crystallographers call it constrained, while they call harmonically constrained particles restrained (Coot) or tethered."
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "import pkg_resources\n",
    "\n",
    "template_block: str = pkg_resources.resource_string('DTC_compchem_practical', 'data/mac1-stripped.pdb')\n",
    "hit_block: str = pkg_resources.resource_string('DTC_compchem_practical', 'data/QRU.mol')\n",
    "hit: Chem.Mol = Chem.MolFromMolBlock(hit_block)\n",
    "\n",
    "# as we are in a rush we will use RDKit's MMFF, in the afternoon we will use OpenMM.\n",
    "# hack: we will use the neighbourhood\n",
    "from fragmenstein import Monster\n",
    "neighborhood = Monster.get_neighborhood(None, template_block, radius=10, mol=hit)\n",
    "\n",
    "# let's see what the neighbourhood looks like\n",
    "view = dtc.get_protein_view(template_block, resn='HOH')\n",
    "dtc.add_mols(view, CyanCarbon=hit, whiteCarbon=neighborhood)\n",
    "view.show()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Get internal energy\n",
    "\n",
    "RDKit has the MMFF94 forcefield (Merck Molecular ForceField).\n",
    "The whole thing can be called to fix a molecules via `AllChem.MMFFOptimizeMolecule`,\n",
    "but often one may want to play with the inner workings.\n",
    "This system is not perfect, but does parameterisation under the hood, which is great for now.\n",
    "Say one has\n",
    "\n",
    "`atom_indices_to_freeze: List[int]` and `atom_indices_to_tether: List[int]`\n",
    "and `system: Chem.Mol`, one can do the following:\n",
    "\n",
    "```python\n",
    "p: AllChem.ForceField.MMFFMolProperties = AllChem.MMFFGetMoleculeProperties(system, 'MMFF94')\n",
    "ff: AllChem.ForceField.ForceField = AllChem.MMFFGetMoleculeForceField(system, p)\n",
    "ff.Initialize()\n",
    "for i in atom_indices_to_freeze:\n",
    "      ff.AddFixedPoint(i)\n",
    "for i in atom_indices_to_tether:\n",
    "      ff.MMFFAddPositionConstraint(i, maxDispl=max_displacement, forceConstant=constraint)\n",
    "outcome: int = ff.Minimize()\n",
    "U: float = ff.CalcEnergy() if outcome == 0 else np.nan\n",
    "```\n",
    "\n",
    "Where `ff.MMFFAddPositionConstraint` adds a bounded harmonic constrain that is zero within `ff_max_displacement`\n",
    "and a `constraint` time the square of the displacement outside of it.\n",
    "\n",
    "So arming ourselves with\n",
    "\n",
    "* the `neighborhood` hack from above\n",
    "* the `ff.CalcEnergy()` function, so we can now calculate the internal energy of the system,\n",
    "* the knowlage that a negative potential is good\n",
    "* the function `AllChem.EmbedMultipleConfs(mol, numConfs=10)` to generate conformers\n",
    "\n",
    "And cheating by using:\n",
    "* the empirical confomer of `hit` from the crystal structure\n",
    "* the function `AllChem.CalcRMS` to get the in place RMSD of the conformers\n",
    "\n",
    "Let's do a crude docking. Doing the following:\n",
    "\n",
    "* Devise a system to keep track of operation and results (e.g. `dict` or `DataClass`)\n",
    "* Cheat by getting the centroid vector of `hit`\n",
    "* Make a copy of `hit` (call it say `vc` for virtual compound)\n",
    "* Generate _n_ conformers of `vc`\n",
    "* make them a `List[Chem.Mol]` (copy `vc`, `copied.RemoveAllConformers`, add a confomer of `vc` to `copied.AddConformer`) ‚Äîthis is a hack, but it is fine for now\n",
    "* Move each conformer to the centroid of `hit`\n",
    "* Combine each mol with `neighborhood` (`AllChem.CombineMols`)\n",
    "* Get the internal energy and store information\n",
    "* Making copies, exhaustively rotating in place by first 90¬∞ around the x-axis, the y-axis, z-axis\n",
    "* See how the best compares to the crystal structure (in place RMSD)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# First we need double check if it's true that the MMFF works in place, and see what happens if we change the conformer after initialisation, which could save us time.\n",
    "# Did the internal energy get worse (positive)?\n",
    "\n",
    "from rdkit.Chem import AllChem\n",
    "from rdkit import Chem, Geometry\n",
    "\n",
    "mol = Chem.MolFromSmiles('CCO')\n",
    "\n",
    "AllChem.EmbedMolecule(mol)\n",
    "\n",
    "p: AllChem.ForceField.MMFFMolProperties = AllChem.MMFFGetMoleculeProperties(mol, 'MMFF94')\n",
    "ff: AllChem.ForceField.ForceField = AllChem.MMFFGetMoleculeForceField(mol, p)\n",
    "ff.Initialize()\n",
    "print('Before... ', ff.CalcEnergy())\n",
    "mol.GetConformer().SetAtomPosition(0, Geometry.Point3D(10,0,0))\n",
    "print('After... ', ff.CalcEnergy())"
   ],
   "metadata": {
    "collapsed": false,
    "is_executing": true
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "Notice anything?\n",
    "\n",
    "The energy was not negative. `AllChem.MMFFOptimizeMoleculeConfs` could be useful here on the initial compound.\n",
    "(Don't do it on the combined neighbourhood as this defeats the purpose of the excercise)\n",
    "\n",
    "RDKit complained: `AllChem.AddHs` (rightfully so!) RDkit and MM in generally will treat impossible chemistry like a triple radical as valid but the partial charges will be wrong along with the atom type allocation.\n",
    "If you know that RDKit is wrong:\n",
    "\n",
    "```python\n",
    "from rdkit import RDLogger\n",
    "logger = RDLogger.logger()\n",
    "logger.setLevel(RDLogger.CRITICAL)\n",
    "```"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Let's do it!\n",
    "\n",
    "Damn. Something went wrong and _you_ will have to do the manual docking.\n",
    "Remember Google, StackOverflow and ChatGTP exists but so does `help` function,\n",
    "and asking!\n",
    "_Enjoy!_"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "üëæüëæüëæ\n",
    "üëæüëæüëæ\n",
    "üëæüëæüëæ"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Concluding question\n",
    "\n",
    "The above was an exhaustive _crude_ search. How would you do a more intelligent search?\n",
    "(cryptic hint: what would Andrey Markov do in a Mon√©gasque casino?)\n",
    "\n",
    "üëæüëæüëæ"
   ],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
